pkg/cloud/services/compute/instance.go

```diff
@@ -407,68 +386,129 @@ func (s *Service) getVolumeByName(name string) (*volumes.Volume, error) {
 	return &volumeList[0], nil
 }
 
-func (s *Service) getOrCreateRootVolume(eventObject runtime.Object, instanceSpec *InstanceSpec, imageID string) (*volumes.Volume, error) {
-	rootVolume := instanceSpec.RootVolume
-	if !hasRootVolume(rootVolume) {
-		return nil, nil
+func (s *Service) getOrCreateVolume(eventObject runtime.Object, opts volumes.CreateOpts) (*volumes.Volume, error) {
+	existingVolume, err := s.getVolumeByName(opts.Name)
+	if err != nil {
+		return nil, err
 	}
+	if existingVolume != nil {
+		// TODO(emilien): Improve the checks here, there is an ongoing discussion in the community about how to do this
+		// which would involve adding metadata to the volume.
+		if existingVolume.Size != opts.Size {
+			return nil, fmt.Errorf("expected to find volume %s with size %d; found size %d", opts.Name, opts.Size, existingVolume.Size)
+		}
 
-	name := rootVolumeName(instanceSpec.Name)
-	size := rootVolume.Size
+		s.scope.Logger().V(3).Info("Using existing volume", "name", opts.Name, "id", existingVolume.ID)
+		return existingVolume, nil
+	}
 
-	volume, err := s.getVolumeByName(name)
+	createdVolume, err := s.getVolumeClient().CreateVolume(opts)
 	if err != nil {
+		record.Eventf(eventObject, "FailedCreateVolume", "Failed to create volume; name=%s size=%d err=%v", opts.Name, opts.Size, err)
 		return nil, err
 	}
-	if volume != nil {
-		if volume.Size != size {
-			return nil, fmt.Errorf("exected to find volume %s with size %d; found size %d", name, size, volume.Size)
+	record.Eventf(eventObject, "SuccessfulCreateVolume", "Created volume; id=%s", createdVolume.ID)
+	return createdVolume, err
+}
+
+func (s *Service) waitForVolume(volumeID string, timeout time.Duration, retryInterval time.Duration) error {
+	return wait.PollUntilContextTimeout(context.TODO(), retryInterval, timeout, true, func(_ context.Context) (bool, error) {
+		volume, err := s.getVolumeClient().GetVolume(volumeID)
+		if err != nil {
+			if capoerrors.IsRetryable(err) {
+				return false, nil
+			}
+			return false, err
 		}
 
-		s.scope.Logger().V(3).Info("Using existing root volume", "name", name)
-		return volume, nil
-	}
+		switch volume.Status {
+		case "available":
+			return true, nil
+		case "error":
+			return false, fmt.Errorf("volume %s is in error state", volumeID)
+		default:
+			return false, nil
+		}
+	})
+}
 
+func (s *Service) getOrCreateVolumeBuilder(eventObject runtime.Object, instanceSpec *InstanceSpec, blockDevice infrav1.AdditionalBlockDevice, imageID string, description string) (*volumes.Volume, error) {
```

Comment:
Please add a doc string for this function

Reply:
done